### ISSUE TITLE
CQB Saving Broken (Repro Included)

### ISSUE THREAD

**Issue by Vegas** [ALiVEOS/ALiVE.OS#390]
Persistence with the CQB module appears to be broken.

I have persistence enabled for CQB, however is does not appear to be saving. After hitting Server Save & Exit, it runs through the save, I see the log entry for CQB, returns OK value, however it always says "saving 1 SQB installation" whatever and never more than 1. After reloading the mission, all CQB is reset and everything is occupied by enemies again.

Also have this report with some more details in the forum here: http://alivemod.com/forum/3057-saving-error-cqb-and-player-logistics-logs

Repro Mission [here](https://mega.nz/#!1hVjjIYJ!VWdQTp8mQ-Z_ad3Y2gqwv0UXvX6c3vXTqDs1beXoKLo)

**Comment by Vegas** [comment]
@highhead @marceldev89

**Comment by HeroesandvillainsOS** [comment]
Hi @Vegas. Ok so I spent some time with your repro (which was perfect for the purposes of this test, so thank you) and can confirm although less than ideal, this is not a bug. CQB is stored not by cleared buildings, but by cleared sectors. Just to give you an example, here is the sector revealed around your mission area so you can see what I'm talking about: 

[img]http://i.imgur.com/dEKBLm4.jpg[/img]

So unless the entirety of sector 12.5 is clear of CQB units, the area won't be considered cleared for persistence purposes, and they will regenerate upon mission load.

That said, I messed around with the mission, and just to clear half of that area, I killed 37 CSAT. And I can certainly see how this would be a buzzkill, spending all that time, saving, and having it revert back. My opinion is that this ticket should be turned into a feature request, so if at all possible, individual CQB positions can be saved with persistence in the future (if at all possible. It may simply store too much data to do it this way). But for now, do your best to clear the entire area. If you want to reveal sectors, enableDebugConsole in your mission and use this code in debug: 

[ALIVE_sectorGrid, "debug", true] call ALiVE_fnc_sectorGrid

**Comment by Vegas** [comment]
@HeroesandvillainsOS Hey, thanks for looking into this for me.

I seem to remember that CQB used to save with a different method, which was much more effective imo. This is ridiculously difficult and time consuming, especially when a "sector" spans many structures in more rural settings as opposed to a city environment where at least things are mostly consolidated.

If saving individual CQB units is not possible, perhaps an alternative would be to at least divide sectors into smaller sections using something similar to a keypad system where a grid is divided into 9 smaller keypad sections.

Nonetheless, thanks for the info and debug command.

**Comment by HeroesandvillainsOS** [comment]
I agree with you. I'd like to see this ticket reverted into a feature request if it's reasonable or possible.

**Comment by Vegas** [comment]
I could honestly see this working several different ways. Here is a little write up.

**The current system**
It appears to be using 4 digit grids, or 1,000m distance to establish sectors. The numbers shown in debug correspond directly with the grid refs. 100% of this sector must be cleared for it to register and save progress.

**Suggested Method 1**
Rewrite the system to simply use smaller grids to establish zones. Using a 6 digit grid would equal 100m distance and would allow for much for detailed map analysis and clearing with the CQB systems. Clearing sections of a city would be possible and players would not be hindered by a single shack on the edge of a town preventing all their work clearing an actual city from saving.

**Suggested Method 2**
Use a keypad system within a grid. IRL grids can be split into smaller areas more easily by using a keypad system, which is taught in most basic land navigation courses. Take a grid, and apply a numbered keypad from 1-9 in rows of 3 across it. This can be read in this example: Grid Ref 123-456, keypad 1, would indicate the North-West upper right corner of that grid. Grid Ref 123-456, keypad 9 would be the bottom right, or South-East.

**Suggested Method 3**
Allow a percentage for error when saving the existing system. This method is no where near as detailed, and could be seen as a cheap fix, but essentially if a player clears 70-80% of a grid, is saves the whole thing as cleared. Obviously the exact percentage could be changed to whatever, but hey, a little room for player error could negate a lot of frustration when a team member fails to clear that one house and you lose all of your city clearing progress.

**Suggested Method 4**
Save every damn CQB location individually. Why not exceed the data limit while we are at it? Assign a numeric ID to every CQB installation and have that save. This would be the most "realistic", but also the most intensive in terms of performance and profiling.


That's about all I can come up with off the top of my head. Without getting into crazy map indexing details these methods seem to be the most viable.

**Comment by friznit** [comment]
Change CQB sector persistence to sub-sector persistence by applying existing sub-sector division func to CQB sectors and persist at sub-sector level (cleared | not cleared)

Add: CQB sector is clear when >75% sub sectors are clear

**Comment by HeroesandvillainsOS** [comment]
@vegas what Friznit is writing is for all intents and purposes going to end up similarly to your first idea in your list. Obviously no promises on implementation but it's being looked at so stay tuned.

**Comment by marceldev89** [comment]
Short update: It seems that CQB persistence didn't work at all even when a full sector was cleared.

Also pretty much implemented method 1, just have to get persistence to work. :)

**Comment by tupolov** [comment]
Pretty sure @highhead had persistence working...

**Comment by marceldev89** [comment]
You're correct. My bad! :heart:

**Comment by marceldev89** [comment]
With the changes above it splits the sector up into 100x100m sub sectors and uses that for persistence. Also added sub sector visualization to CQB's debug.

**Comment by Vegas** [comment]
You guys are doing an outstanding job. Thank you all for your hard work and diligent response to feedback.
